railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	// check if CR current status already updated
+	var status []apiv1.LoadBalancerIngress
+	sort.SliceStable(status, lessLoadBalancerIngress(status))
+	curIPs := toCoreLBStatus(curIng.Status.PublicLoadBalancer)
+	ips, err := RunningAddresses(ctx, kubeCfg)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster loadbalancer.")
+	}
+	status = SliceToStatus(ips)
+	if IngressSliceEqual(status, curIPs) &&
+		curIng.Status.ObservedGeneration == curIng.GetObjectMeta().GetGeneration() {
+		log.Info("no change in status, update skipped")
+		return nil
+	}
+
+	// updating current custom status
+	log.Info("attempting to update Knative Ingress status")
+	lbStatus := toKnativeLBStatus(status)
+	clusterDomain := network.GetClusterDomainName()
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(lbStatus); i++ {
+		lbStatus[i].DomainInternal = fmt.Sprintf("%s.%s.svc.%s",
+			"ingress-controller-kong-proxy", "kong-system", clusterDomain)
+	}
+
+	curIng.Status.MarkLoadBalancerReady(lbStatus, lbStatus)
+	ingressCondSet.Manage(&curIng.Status).MarkTrue(knative.IngressConditionReady)
+	ingressCondSet.Manage(&curIng.Status).MarkTrue(knative.IngressConditionNetworkConfigured)
+	curIng.Status.ObservedGeneration = curIng.GetObjectMeta().GetGeneration()
+
+	_, err = ingClient.UpdateStatus(ctx, curIng, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update ingress status: %v", err)
+	} else {
+		logger.Info("successfully updated ingress status")
+	}
+	return nil
+}
+
+func RunningAddresses(ctx context.Context, kubeCfg *rest.Config) ([]string, error) {
+	addrs := []string{}
+
+	namespace := "kong-system"
+	CoreClient, _ := clientset.NewForConfig(kubeCfg)
+	svc, err := CoreClient.CoreV1().Services(namespace).Get(ctx, "ingress-controller-kong-proxy", metav1.GetOptions{})
+	if err != nil {
+		log.Infof("running address err %v", err)
+		return nil, err
+	}
+
+	switch svc.Spec.Type {
+	case apiv1.ServiceTypeLoadBalancer:
+		for _, ip := range svc.Status.LoadBalancer.Ingress {
+			if ip.IP == "" {
+				addrs = append(addrs, ip.Hostname)
+			} else {
+				addrs = append(addrs, ip.IP)
+			}
+		}
+
+		addrs = append(addrs, svc.Spec.ExternalIPs...)
+		return addrs, nil
+	default:
+		return addrs, nil
+	}
+}
+
+// sliceToStatus converts a slice of IP and/or hostnames to LoadBalancerIngress
+func SliceToStatus(endpoints []string) []apiv1.LoadBalancerIngress {
+	lbi := []apiv1.LoadBalancerIngress{}
+	for _, ep := range endpoints {
+		if net.ParseIP(ep) == nil {
+			lbi = append(lbi, apiv1.LoadBalancerIngress{Hostname: ep})
+		} else {
+			lbi = append(lbi, apiv1.LoadBalancerIngress{IP: ep})
+		}
+	}
+
+	sort.SliceStable(lbi, func(a, b int) bool {
+		return lbi[a].IP < lbi[b].IP
+	})
+
+	return lbi
+}
+
+func InSlice(e string, arr []string) bool {
+	for _, v := range arr {
+		if v == e {
+			return true
+		}
+	}
+
+	return false
+}
+
+func IngressSliceEqual(lhs, rhs []apiv1.LoadBalancerIngress) bool {
```

Comment:
lint: public symbol needs godoc comment (we don't seem to have our linter set up properly? it should be failing CI for these)